File: acm/uri/1455-icpc-finals.py
#!/usr/bin/env python3

import random
from itertools import count
from collections import namedtuple


Point = namedtuple('Point', 'x y')
Point.from_spec = lambda st: Point(*[float(si) for si in st.split()])

Circle = namedtuple('Circle', 'c r')
Circle.__new__.__defaults__ = (Point(0, 0), 0)
Circle.__contains__ = lambda s, p: s.r > 0 and sq_dist(p, s.c) <= s.r**2
Circle.from_arc = lambda *ps: ( circle_from_1_points(*ps) if len(ps) == 1 else
                                circle_from_2_points(*ps) if len(ps) == 2 else
                                circle_decide_3_points(*ps) if len(ps) == 3 else
                                NotImplemented )

sq_dist = lambda p, q=Point(0, 0): (p.x-q.x)**2 + (p.y-q.y)**2

det = lambda u, v, w: u.x*v.y + v.x*w.y + w.x*u.y - u.x*w.y - v.x*u.y - w.x*v.y
detx = lambda u, v, w: ( sq_dist(u)*v.y + sq_dist(v)*w.y + sq_dist(w)*u.y
                       - sq_dist(u)*w.y - sq_dist(v)*u.y - sq_dist(w)*v.y )
dety = lambda u, v, w: ( u.x*sq_dist(v) + v.x*sq_dist(w) + w.x*sq_dist(u)
                       - u.x*sq_dist(w) - v.x*sq_dist(u) - w.x*sq_dist(v) )
dets = lambda u, v, w: ( u.x*v.y*sq_dist(w) + v.x*w.y*sq_dist(u) + w.x*u.y*sq_dist(v)
                       - u.x*w.y*sq_dist(v) - v.x*u.y*sq_dist(w) - w.x*v.y*sq_dist(u) )


def shuffled(points):
    ls = list(points)
    random.shuffle(ls)
    return ls


# XXX please note that though the problem statement says 2 <= n <= 100,
#     the actual test does contain case of n == 1 too. WTF!!!
def circle_from_1_points(p1):
    return Circle(p1, 0)


def circle_from_2_points(p1, p2):
    dx = (p1.x + p2.x)/2
    dy = (p1.y + p2.y)/2
    c = Point(dx, dy)
    r = sq_dist(p1, p2)**.5/2
    return Circle(c, r)


def circle_from_3_points(p1, p2, p3):
    a = det(p1, p2, p3)
    b = dets(p1, p2, p3)
    sx = detx(p1, p2, p3)
    sy = dety(p1, p2, p3)
    c = Point(sx/a/2, sy/a/2)
    return Circle(c, (b/a + sq_dist(c))**.5)


def circle_decide_3_points(p1, p2, p3):
    ps = {p1, p2, p3}
    for p in ps:
        circle = circle_from_2_points(*ps-{p})
        if p in circle:
            return circle
    return circle_from_3_points(p1, p2, p3)


def mincircle2p(points, p1, p2):
    circle = Circle.from_arc(p1, p2)
    for i, p in enumerate(points):
        if p not in circle:
            circle = Circle.from_arc(p1, p2, p)
    return circle


def mincircle1p(points, p1):
    circle = Circle.from_arc(p1)
    ps = shuffled(points)
    for i, p in enumerate(ps):
        if p not in circle:
            circle = mincircle2p(set(ps[:i]), p1, p)
    return circle


def mincircle(points):
    circle = Circle()
    ps = shuffled(points)
    for i, p in enumerate(ps):
        if p not in circle:
            circle = mincircle1p(set(ps[:i]), p)
    return circle


def main():
    for t in count(1):
        n = int(input())
        if n == 0:
            break
        points = [Point.from_spec(input()) for _ in range(n)]
        ans = mincircle(points)
        print('Instancia {}'.format(t))
        print('{:.2f} {:.2f} {:.2f}'.format(ans.c.x, ans.c.y, ans.r))
        print()


if __name__ == '__main__':
    main()

File: acm/uri/1455-icpc-finals.cpp
#include <bits/stdc++.h>
using namespace std;

typedef pair<double,double> pt;
typedef pair<pt,double> circ;


unsigned get_seed() {
    return chrono::system_clock::now().time_since_epoch().count();
}


double det(double m11, double m12, double m13,
           double m21, double m22, double m23,
           double m31, double m32, double m33) {
    return m11*m22*m33 + m12*m23*m31 + m13*m21*m32 - m31*m22*m13 - m32*m23*m11 - m33*m21*m12;
}


pt center(circ circle) {
    return circle.first;
}


double sq_dist(pt p1, pt p2=make_pair(0, 0)) {
    return pow(p1.first-p2.first, 2.0) + pow(p1.second-p2.second, 2.0);
}


// XXX please note that though the problem statement says 2 <= n <= 100,
//     the actual test does contain case of n == 1 too. WTF!!!
circ from_1_points(pt p1) {
    return make_pair(p1, 0);
}


circ from_2_points(pt p1, pt p2) {
    double dx = (p1.first + p2.first) / 2.0;
    double dy = (p1.second + p2.second) / 2.0;
    pt c = make_pair(dx, dy);
    double r = pow(sq_dist(p1, p2), 0.5) / 2.0;
    return make_pair(c, r);
}


circ from_3_points(pt p1, pt p2, pt p3) {
    double ss1 = sq_dist(p1);
    double ss2 = sq_dist(p2);
    double ss3 = sq_dist(p3);
    double a = det(p1.first, p1.second, 1, p2.first, p2.second, 1, p3.first, p3.second, 1);
    double b = det(p1.first, p1.second, ss1, p2.first, p2.second, ss2, p3.first, p3.second, ss3);
    double sx = det(ss1, p1.second, 1, ss2, p2.second, 1, ss3, p3.second, 1);
    double sy = det(p1.first, ss1, 1, p2.first, ss2, 1, p3.first, ss3, 1);
    pt c = make_pair(sx/a/2.0, sy/a/2.0);
    double r = pow(b/a + sq_dist(c), 0.5);
    return make_pair(c, r);
}


bool contains(circ circle, pt p) {
    if (circle.second == 0) {
        return false;
    }
    return sq_dist(p, center(circle)) <= pow(circle.second, 2.0);
}


circ mincircle3p(pt p1, pt p2, pt p3) {
    if (contains(from_2_points(p1, p3), p2)) {
        return from_2_points(p1, p3);
    }
    if (contains(from_2_points(p2, p3), p1)) {
        return from_2_points(p2, p3);
    }
    return from_3_points(p1, p2, p3);
}


circ mincircle2p(vector<pt> ps, pt p1, pt p2) {
    circ circle = from_2_points(p1, p2);
    for (int i=0; i<size(ps); i++) {
        if (!contains(circle, ps[i])) {
            circle = mincircle3p(p1, p2, ps[i]);
        }
    }
    return circle;
}


circ mincircle1p(vector<pt> ps, pt p1) {
    vector<pt> seen;
    shuffle(ps.begin(), ps.end(), default_random_engine(get_seed()));
    circ circle = from_1_points(p1);
    for (int i=0; i<size(ps); i++) {
        if (!contains(circle, ps[i])) {
            circle = mincircle2p(seen, p1, ps[i]);
        }
        seen.push_back(ps[i]);
    }
    return circle;
}


circ mincircle(vector<pt> ps) {
    vector<pt> seen;
    shuffle(ps.begin(), ps.end(), default_random_engine(get_seed()));
    circ circle = make_pair(make_pair(0, 0), 0);
    for (int i=0; i<size(ps); i++) {
        if (!contains(circle, ps[i])) {
            circle = mincircle1p(seen, ps[i]);
        }
        seen.push_back(ps[i]);
    }
    return circle;
}


int main(void) {
    int n;
    for (int t=1;;t++) {
        cin >> n;
        if (n == 0) {
            break;
        }
        vector<pt> ps;
        double x, y;
        for (int i=0; i<n; i++) {
            cin >> x >> y;
            ps.push_back(make_pair(x, y));
        }
        circ ans = mincircle(ps);
        printf("Instancia %d\n", t);
        printf("%.2f %.2f %.2f\n\n", ans.first.first, ans.first.second, ans.second);
    }
    return 0;
}
